### PR TITLE
Fix cond batching rule issues

### LINKED ISSUE #1263: cond batching rule shape mismatch
A simple script to repro
```
from jax import lax, numpy as np, vmap, jit, random
from jax.config import config; config.update('jax_platform_name', 'cpu')

def f(rng, x):
    cond = random.bernoulli(rng)
    y = np.abs(x) - 1.
    return lax.cond(cond, x, lambda x: x, y, lambda x: x)

def body_fn(i, state):
    rng, x = state
    key, subkey = random.split(rng)
    return key, f(subkey, x)

def g(rng, x):
    return lax.fori_loop(0, 10, body_fn, (rng, x))

vmap(g)(random.split(random.PRNGKey(0), 3), np.ones((3, 4)))
```
which triggers
```
TypeError: select pred must be scalar or have the same shape as on_true and on_false, got pred shape (3,) for on_true and on_false of shape (3, 4).
```
More details

<details>

```sh
~/jax/jax/lax/lax_control_flow.py in fori_loop(lower, upper, body_fun, init_val)
    116     return lax.add(i, lax._const(i, 1)), body_fun(i, x)
    117 
--> 118   _, result = while_loop(while_cond_fun, while_body_fun, (lower, init_val))
    119   return result
    120 

~/jax/jax/lax/lax_control_flow.py in while_loop(cond_fun, body_fun, init_val)
    171   outs = while_p.bind(*itertools.chain(cond_consts, body_consts, init_vals),
    172                       cond_nconsts=len(cond_consts), cond_jaxpr=cond_jaxpr,
--> 173                       body_nconsts=len(body_consts), body_jaxpr=body_jaxpr)
    174   return tree_unflatten(body_tree, outs)
    175 

~/jax/jax/core.py in bind(self, *args, **kwargs)
    131 
    132     tracers = map(top_trace.full_raise, args)
--> 133     out_tracer = top_trace.process_primitive(self, tracers, kwargs)
    134     if self.multiple_results:
    135       return map(full_lower, out_tracer)

~/jax/jax/interpreters/batching.py in process_primitive(self, primitive, tracers, params)
    113       # TODO(mattjj,phawkins): if no rule implemented, could vmap-via-map here
    114       batched_primitive = get_primitive_batcher(primitive)
--> 115       val_out, dim_out = batched_primitive(vals_in, dims_in, **params)
    116       if primitive.multiple_results:
    117         return map(partial(BatchTracer, self), val_out, dim_out)

~/jax/jax/lax/lax_control_flow.py in _while_loop_batching_rule(args, dims, cond_nconsts, cond_jaxpr, body_nconsts, body_jaxpr)
    247     batched = bconst_bat + carry_bat
    248     body_jaxpr_batched, carry_bat_out = batching.batch_jaxpr(
--> 249         body_jaxpr, size, batched, instantiate=carry_bat)
    250     cond_jaxpr_batched, (pred_bat,) = batching.batch_jaxpr(
    251         cond_jaxpr, size, cconst_bat + carry_bat, instantiate=False)

~/jax/jax/interpreters/batching.py in batch_jaxpr(jaxpr, size, batched, instantiate)
    293               for a, b in zip(jaxpr.in_avals, batched)]
    294   in_pvals = [pe.PartialVal((aval, core.unit)) for aval in avals_in]
--> 295   jaxpr_out, pvals_out, consts_out = pe.trace_to_jaxpr(f, in_pvals, instantiate=True)
    296   avals_out, _ = unzip2(pvals_out)
    297   jaxpr_out = core.TypedJaxpr(jaxpr_out, consts_out, avals_in, avals_out)

~/jax/jax/interpreters/partial_eval.py in trace_to_jaxpr(fun, pvals, **kwargs)
    316   with new_master(JaxprTrace) as master:
    317     fun = trace_to_subjaxpr(fun, master, instantiate)
--> 318     jaxpr, (out_pvals, consts, env) = fun.call_wrapped(pvals)
    319     assert not env
    320     del master

~/jax/jax/linear_util.py in call_wrapped(self, *args, **kwargs)
    163 
    164     del gen
--> 165     ans = self.f(*args, **dict(self.params, **kwargs))
    166     del args
    167     while stack:

~/jax/jax/core.py in jaxpr_as_fun(typed_jaxpr, *args)
     77 @curry
     78 def jaxpr_as_fun(typed_jaxpr, *args):
---> 79   return eval_jaxpr(typed_jaxpr.jaxpr, typed_jaxpr.literals, (), *args)
     80 
     81 

~/jax/jax/core.py in eval_jaxpr(jaxpr, consts, freevar_vals, *args)
    183                in eqn.bound_subjaxprs]
    184     subfuns = map(lu.wrap_init, subfuns)
--> 185     ans = eqn.primitive.bind(*(subfuns + in_vals), **eqn.params)
    186     if eqn.primitive.multiple_results:
    187       map(write, eqn.outvars, ans)

~/jax/jax/core.py in bind(self, *args, **kwargs)
    131 
    132     tracers = map(top_trace.full_raise, args)
--> 133     out_tracer = top_trace.process_primitive(self, tracers, kwargs)
    134     if self.multiple_results:
    135       return map(full_lower, out_tracer)

~/jax/jax/interpreters/batching.py in process_primitive(self, primitive, tracers, params)
    113       # TODO(mattjj,phawkins): if no rule implemented, could vmap-via-map here
    114       batched_primitive = get_primitive_batcher(primitive)
--> 115       val_out, dim_out = batched_primitive(vals_in, dims_in, **params)
    116       if primitive.multiple_results:
    117         return map(partial(BatchTracer, self), val_out, dim_out)

~/jax/jax/lax/lax_control_flow.py in _cond_batching_rule(args, dims, true_jaxpr, false_jaxpr, true_nconsts, false_nconsts)
    358                 for x, b in zip(false_out, out_bat)]
    359     return [lax.select(pred, t, f)
--> 360             for t, f in zip(true_out, false_out)], [0] * len(true_out)
    361   else:
    362     out_dims = [0 if b else batching.not_mapped for b in out_bat]

~/jax/jax/lax/lax_control_flow.py in <listcomp>(.0)
    358                 for x, b in zip(false_out, out_bat)]
    359     return [lax.select(pred, t, f)
--> 360             for t, f in zip(true_out, false_out)], [0] * len(true_out)
    361   else:
    362     out_dims = [0 if b else batching.not_mapped for b in out_bat]

~/jax/jax/lax/lax.py in select(pred, on_true, on_false)
    611   operator.
    612   """
--> 613   return select_p.bind(pred, on_true, on_false)
    614 
    615 def slice(operand, start_indices, limit_indices, strides=None):

~/jax/jax/core.py in bind(self, *args, **kwargs)
    131 
    132     tracers = map(top_trace.full_raise, args)
--> 133     out_tracer = top_trace.process_primitive(self, tracers, kwargs)
    134     if self.multiple_results:
    135       return map(full_lower, out_tracer)

~/jax/jax/interpreters/partial_eval.py in process_primitive(self, primitive, tracers, params)
     93       tracers = map(self.instantiate_const, tracers)
     94       avals = [t.aval for t in tracers]
---> 95       out_aval = primitive.abstract_eval(*avals, **params)
     96       # TODO(dougalm): think about whether these ref cycles will leak memory
     97       if primitive.multiple_results:

~/jax/jax/lax/lax.py in standard_abstract_eval(shape_rule, dtype_rule, *args, **kwargs)
   1468     return ShapedArray(shape_rule(*args, **kwargs), dtype_rule(*args, **kwargs))
   1469   elif least_specialized is ShapedArray:
-> 1470     return ShapedArray(shape_rule(*args, **kwargs), dtype_rule(*args, **kwargs))
   1471   elif least_specialized is UnshapedArray:
   1472     return UnshapedArray(dtype_rule(*args, **kwargs))

~/jax/jax/lax/lax.py in _select_shape_rule(pred, on_true, on_false)
   2561     msg = ("select pred must be scalar or have the same shape as on_true and "
   2562            "on_false, got pred shape {} for on_true and on_false of shape {}.")
-> 2563     raise TypeError(msg.format(pred.shape, on_true.shape))
   2564   return on_true.shape
   2565 
```
  </details>

I think that the logic of `cond_batching_rule` is correct but the `pred` shape is somehow not broadcasted in `_cond_batching_rule`.
cc @jekbradbury @mattjj @neerajprad

### PR DESCRIPTION
**Pull request by fehiepsi**
Fixes #1263, which lacks broadcasting mechanism for `select` (thanks @mattjj)

Fixes a swap of `tconst_bat, t_bat, fconst_bat, f_bat` in batching rule implementation too.

### CODE REVIEW

**Comment by mattjj** [comment]
Thanks for catching this, fixing it, and adding tests! Beautiful!

Unfortunately we can't depend on `jax.numpy` inside lax; we want `jax.numpy` to be 'downstream' of all the lax stuff, as a means of avoiding circular imports. So the broadcasting needs to be done with `lax.broadcast_in_dim`, which is kind of a pain but isn't too bad.

WDYT?

**Comment by fehiepsi** [comment]
Yeah, that makes sense to me. I looked at the corresponding one at while_loop but it is implemented using XLA language, so I decided to stick with numpy. I will find the corresponding ones in `lax` instead. Thanks for pointing out `lax.broadcast_in_dim`!

File: jax/lax/lax_control_flow.py [hunk 1]
```diff
@@ -329,6 +329,11 @@ def make_computation(name, jaxpr, op_shape):
 
   return c.Conditional(pred, true_op, true_c, false_op, false_c)
 
+def _cond_pred_bcast_select(pred, x, y):
+  promote_pred = lax.reshape(pred, onp.shape(pred) + (1,) * (x.ndim - pred.ndim))
```
Comment (mattjj) [inline jax/lax/lax_control_flow.py]: I think you can avoid the reshape and instead write this as

```python
bcast_pred = lax.broadcast_in_dim(pred, onp.shape(x), list(range(onp.ndim(pred))))
```

File: jax/lax/lax_control_flow.py [hunk 2]
```diff
@@ -339,7 +344,7 @@ def _cond_batching_rule(args, dims, true_jaxpr, false_jaxpr, true_nconsts,
       args, [1, true_nconsts, true_nops, false_nconsts])
   size, = {x.shape[d] for x, d in zip(args, dims) if d is not batching.not_mapped}
   orig_bat = [d is not batching.not_mapped for d in dims]
-  (pred_bat,), t_bat, tconst_bat, f_bat, fconst_bat = split_list(
+  (pred_bat,), tconst_bat, t_bat, fconst_bat, f_bat = split_list(
```
Comment (mattjj) [inline jax/lax/lax_control_flow.py]: Should this be `true_nops` instead of `len(true_ops)`, following the `split_list(args, ...)` call above?

**Review by mattjj — APPROVED** [review]
Nice! A couple comments (one a simplification, the other catching what I think is a bug that existed before this PR).

File: jax/lax/lax_control_flow.py (hunk 2 above)
Comment (fehiepsi) [reply jax/lax/lax_control_flow.py]: I guess they are the same but using `true_nops` is clearer to me.

Comment (mattjj) [reply jax/lax/lax_control_flow.py]: Ah, you're right, this isn't a bug: I was getting confused with `len(true_jaxpr.in_avals)`. But `true_nops == len(true_ops)` of course!

Either version is good with me; one nice thing about writing `true_nops` is that it matches the unpacking done above, which makes it more obviously correct.

File: jax/lax/lax_control_flow.py (hunk 1 above)
Comment (fehiepsi) [reply jax/lax/lax_control_flow.py]: done

**Comment by mattjj** [comment]
Woohoo! I'll merge when the tests finish!

**Comment by mattjj** [comment]
Thanks for this!

**Pull request merged**